### DISCUSSION TITLE
panic: runtime error: slice bounds out of range [-51:] while rows.Next()

### DISCUSSION THREAD

**Discussion by adamszpilewicz** [sijms/go-ora#357] [Q&A]
I am testing now `go-ora` library to mine LogMiner contents view.
And while using I play with different number of PREFETCH_ROWS options. And in case of setting it to 150 or 200 I often have panic like panic: runtime error: slice bounds out of range [-51:] that comes from for `rows.Next()`
Do you have any idea why it could happen? Can it be related to PREFETCH_ROWS options, that I should decrease?

```go
	// Define URL options
	urloptions := map[string]string{
		"PREFETCH_ROWS": strconv.Itoa(prefetchRows),
	}

	// Create the connection string using go-ora.BuildUrl
	connStr := goora.BuildUrl(host, port, serviceName, username, password, urloptions)
```

```go
// ParseLogMinerContents parses the contents of the logminer view and returns a slice of LogMinerContent.
func (lm *Logminer) ParseLogMinerContents(rows *sql.Rows, tableName string) ([]LogMinerContent, error) {
	var contents []LogMinerContent
	var hasRows bool

	var i int
	for rows.Next() {
		hasRows = true
		i++
		if i%1000 == 0 {
			log.Printf("Parsed rows: %d for table: %s\n", i, tableName)
		}
		var content LogMinerContent
		var xidHex string // Temporary variable to hold the hexadecimal string

		err := rows.Scan(
			&content.SCN,
			&content.SQLRedo,
			&content.OperationCode,
			&content.Timestamp,
			&xidHex,
			&content.CSF,
			&content.TableName,
			&content.SegOwner,
			&content.Operation,
			&content.Username,
			&content.RowID,
			&content.RedoValue,
			&content.Rollback,
			&content.RsID,
			&content.Status,
			&content.Info,
			&content.SSN,
			&content.ThreadNum,
		)
		if err != nil {
			return nil, fmt.Errorf("error scanning row: %v", err)
		}

		content.XID = fmt.Sprintf("%X", xidHex) // Convert the hexadecimal string to bytes

		contents = append(contents, content)
	}

	if rows.Err() != nil {
		return nil, fmt.Errorf("error reading rows: %v", rows.Err())
	}

	if hasRows {
		log.Printf("LogMiner contents parsed for the table: %s\n", tableName)
	}
	return contents, nil
}
```

and the whole panic:
```bash
panic: runtime error: slice bounds out of range [-51:]

goroutine 1 [running]:
github.com/sijms/go-ora/v2/network.(*Session).GetInt64(0x14001260f80?, 0x14000bc5c9b?, 0x19?, 0x2?)
        /Users/adamszpilewicz/go/pkg/mod/github.com/sijms/go-ora/v2@v2.7.3/network/session.go:1087 +0x184
github.com/sijms/go-ora/v2/network.(*Session).GetInt(...)
        /Users/adamszpilewicz/go/pkg/mod/github.com/sijms/go-ora/v2@v2.7.3/network/session.go:1103
github.com/sijms/go-ora/v2/network.(*Session).GetClr(0x1400018a1c0)
        /Users/adamszpilewicz/go/pkg/mod/github.com/sijms/go-ora/v2@v2.7.3/network/session.go:1145 +0x19c
github.com/sijms/go-ora/v2.(*ParameterInfo).decodeValue(0x14000eac0f0, 0x1400013b680)
        /Users/adamszpilewicz/go/pkg/mod/github.com/sijms/go-ora/v2@v2.7.3/parameter.go:886 +0xf4
github.com/sijms/go-ora/v2.(*ParameterInfo).decodeColumnValue(0x14000eac0f0, 0xcfce10?)
        /Users/adamszpilewicz/go/pkg/mod/github.com/sijms/go-ora/v2@v2.7.3/parameter.go:1072 +0x388
github.com/sijms/go-ora/v2.(*defaultStmt).calculateColumnValue(0x14001550e60, 0x14000eac0f0)
        /Users/adamszpilewicz/go/pkg/mod/github.com/sijms/go-ora/v2@v2.7.3/command.go:1482 +0x540
github.com/sijms/go-ora/v2.(*defaultStmt).read(0x14001550e60, 0x1400006c240)
        /Users/adamszpilewicz/go/pkg/mod/github.com/sijms/go-ora/v2@v2.7.3/command.go:784 +0xe7c
github.com/sijms/go-ora/v2.(*defaultStmt)._fetch(0x14001550e60, 0x1059d7780?)
        /Users/adamszpilewicz/go/pkg/mod/github.com/sijms/go-ora/v2@v2.7.3/command.go:634 +0x148
github.com/sijms/go-ora/v2.(*defaultStmt).fetch(0x14000066000?, 0x1057234c0?)
        /Users/adamszpilewicz/go/pkg/mod/github.com/sijms/go-ora/v2@v2.7.3/command.go:577 +0x5c
github.com/sijms/go-ora/v2.(*DataSet).Next(0x1400006c240, {0x1400026f440, 0x12, 0x200?})
        /Users/adamszpilewicz/go/pkg/mod/github.com/sijms/go-ora/v2@v2.7.3/data_set.go:310 +0x1c4
database/sql.(*Rows).nextLocked(0x14001666000)
        /opt/homebrew/Cellar/go/1.20.1/libexec/src/database/sql/sql.go:2974 +0x158
database/sql.(*Rows).Next.func1()
        /opt/homebrew/Cellar/go/1.20.1/libexec/src/database/sql/sql.go:2952 +0x30
database/sql.withLock({0x10579a960, 0x14001666030}, 0x14001261918)
        /opt/homebrew/Cellar/go/1.20.1/libexec/src/database/sql/sql.go:3405 +0x7c
database/sql.(*Rows).Next(0x14001666000)
        /opt/homebrew/Cellar/go/1.20.1/libexec/src/database/sql/sql.go:2951 +0x64
orac/logminer.(*Logminer).ParseLogMinerContents(0x14000076460?, 0x0?, {0x1052da14a, 0xe})

```

**Comment by sijms** [comment]
OK i will test the code

**Comment by sijms** [comment]
would you pls send sql text

**Reply by adamszpilewicz** [reply]
Let me try to reproduce, as I made one change in my code and currently do not have that issue even when I set PREFETCHED_ROWS to 450.

The query is the same, as I was reading logminer contents view, but previosuly I was reading from active logs. Now I corrected code in my app to read only from archived logs, that should have only commited transactions. Therefore I suppose it could be problem with my first approch, when I was reading the active logs and long transaction were aborted during doing such operations as below.

I will test that and come with the answer whether still have that panic.

Operation I was conducting on db and tested its cancellation.
```sql
DECLARE
    i            NUMBER := 1;
    random_value NUMBER;
BEGIN
    WHILE i <= 50000
        LOOP
            random_value := i + DBMS_RANDOM.VALUE;
            INSERT INTO RIVERY.MANY_TYPES_CDC (MY_FLOAT, MY_NUMBER) VALUES (random_value, i);
            UPDATE RIVERY.MANY_TYPES_CDC SET MY_FLOAT = random_value - 1 WHERE MY_NUMBER = i;
            DELETE FROM RIVERY.MANY_TYPES_CDC WHERE MY_NUMBER = i;
            i := i + 1;
        END LOOP;
    commit;
END;
/
```

And my query to read that was:
```sql
SELECT SCN, SQL_REDO, OPERATION_CODE, TIMESTAMP, XID, CSF, TABLE_NAME, SEG_OWNER, 
       OPERATION, USERNAME, ROW_ID,REDO_VALUE, ROLLBACK, RS_ID, STATUS, INFO, SSN, THREAD# AS THREAD
        FROM V$LOGMNR_CONTENTS
        WHERE SEG_OWNER = :1
          AND TABLE_NAME = :2
          AND SCN > :3 AND SCN <= :4
          AND OPERATION_CODE in (1, 3, 2)

```

**Comment by adamszpilewicz** [comment]
Unfortunartely issue still persists.

I have this panic or hanging while fetching records from database.
I do think it is related to `PREFETCH_ROWS` I set up when starting the connection. As in my case the default 25 is much too low - it works but it takes too much time to get data from database. My app should replicate changes near real time, therefore I would like to have `PREFETCH_ROWS` set to big number - 2000 or even more.

Is it the limitation of that library? I mean setting up the `PREFETCH_ROWS` to huge numbers to be able to get data fast in a loop like:
```go
for rows.Next() {
```

![image](https://github.com/sijms/go-ora/assets/57848391/fe0baa73-eae2-4145-b5da-2e0590d14d0a)

**Comment by sijms** [comment]
limitation is related to network. so if you have slow network don't use high value of PREFETCH rows. use value < 500
if you have high speed network you can use 2000 or more

**Reply by adamszpilewicz** [reply]
Many thanks for the answer.
And that panic error that I have when setting high prefetched rows can be somehow related to logminer view that I query?
As increasing that prefetched_rows my app is more likely to panic like that.
Should I provide more details?

https://docs.oracle.com/en/database/oracle/oracle-database/19/sutil/oracle-logminer-utility.html#GUID-14D0D046-79B1-4A3F-8F6A-DC98598D2B7D

**Comment by sijms** [comment]
do you work on local network or internet?
I think this error is related to bug in driver when reading some information from logminer view

**Reply by adamszpilewicz** [reply]
Over internet, my setup is:
- I am using ssh tunnel to connect to database
- then connecting to it: Opening connection to db using go-ora: `oracle://admin:[REDACTED]@localhost:2525/ORACLE?PREFETCH_ROWS=5000`

**Comment by sijms** [comment]
I think network like this is not stable and 5000 is large number reduce it to 500-1000

**Reply by adamszpilewicz** [reply]
Reducing to 1000 as well sometimes leads to that error. When I was debugging I see in that part of the code for example.
Is it something that can be fixed?

![image](https://github.com/sijms/go-ora/assets/57848391/8565b19b-dd0f-415c-aae0-ce5d0f8c0027)

**Comment by sijms** [comment]
error in read(). read is a very big function that has many parts
if the error repeated during query a specific sql stmt then we can follow and debug
but it error occur occasionally it is difficult
please review issue #112 which discuss the relation between prefetch rows and network stability

**Reply by adamszpilewicz** [reply]
I see that I got that error here in the `copy` part of the function:

```bash
2023/05/13 17:32:28 Parsed rows: 30000 for table: SRC_FIELDS_2
panic: runtime error: slice bounds out of range [-31:]

goroutine 1 [running]:
github.com/sijms/go-ora/v2/network.(*Session).GetInt64(0x140001141c0, 0x27, 0x1, 0x1)
        /Users/adamszpilewicz/go/pkg/mod/github.com/sijms/go-ora/v2@v2.7.3/network/session.go:1087 +0x404
github.com/sijms/go-ora/v2/network.(*Session).GetInt(0x140001141c0, 0x4, 0x1, 0x1)
        /Users/adamszpilewicz/go/pkg/mod/github.com/sijms/go-ora/v2@v2.7.3/network/session.go:1103 +0x4c
github.com/sijms/go-ora/v2/network.(*Session).GetClr(0x140001141c0)
        /Users/adamszpilewicz/go/pkg/mod/github.com/sijms/go-ora/v2@v2.7.3/network/session.go:1145 +0x24c

```

![image](https://github.com/sijms/go-ora/assets/57848391/80cd5f7f-3803-4381-9b19-c8c58cdac326)

**Comment by sijms** [comment]
coy fail because size is out of range and size is read from the stream.
so this means stream has issue and you can confirm that by switch on data integrity service from the server

**Reply by adamszpilewicz** [reply]
Ok really thanks for the response.
I have tried with the Prepared statement, but with the same result.
Do you know whether there can be some solution other than decreasing prefetched rows count?
As I would like to have better permormance and the prefetched_rows lower than 500 does not meet that requirement?

```go
// GetLogMinerContents returns the contents of the logminer session.
// It has to be called after DBMS_LOGMNR.START_LOGMNR.
func (lm *Logminer) GetLogMinerContents(v_last_scn int64, v_new_scn int64, segOwner string, tableName string) (*sql.Rows, error) {
	//operationCodes := fmt.Sprintf("%d, %d, %d", INSERT_CODE, UPDATE_CODE, DELETE_CODE)
	// startScn is the exclusive lower bound, so must be (v_last_scn - 1)
	log.Printf("Quering logminer with parameters: segOwner: %s, tableName: %s, v_last_scn: %d, v_new_scn: %d\n", segOwner, tableName, v_last_scn, v_new_scn)
	stmt, err := lm.DB.Prepare(queryGetLogminerContents)
	if err != nil {
		return nil, fmt.Errorf("error preparing statement: %w", err)
	}
	defer func(stmt *sql.Stmt) {
		err := stmt.Close()
		if err != nil {
			log.Printf("error closing statement: %v", err)
		}
	}(stmt)
	rows, err := stmt.Query(segOwner, tableName, v_last_scn, v_new_scn)
	return rows, err
}
```

**Comment by sijms** [comment]
ok no problem I will follow the first assumption ( there is a bug in the driver)
if you catch the error would you share trace.log file so I can analyze and see where is the bug

**Comment by sijms** [comment]
also ensure that you use latest version

**Reply by adamszpilewicz** [reply]
Yes I am using the last one.
`trace.log` sent on your mail.
Thanks.
Cheers

**Comment by sijms** [comment]
ok. I mean oracle network trace you can create this file simply by pass a url options like this
```golang
urlOptions := map[string] string {
    "trace file" : "trace.log",
}
connStr := go_ora.BuildUrl(server, port, service, user, password, urlOptions)
conn, err = sql.Open("oracle", connStr)
// rest of your code
```
if you open the file you will see sections the section that is important for me
1- advance negotiation if your connection using encryption or data integrity services other wise not important
2- server charset and server national charset
3- database version
4- all sql data read/write until end of the file (where error happen)
